### PR TITLE
Support for stashing sensitive variables on Rpc and EventHandler

### PR DESCRIPTION
**Pull request by mattbennett**
entrypoints

### CODE REVIEW

File: nameko/events.py [hunk 1]
```diff
@@ -101,64 +101,77 @@ def dispatch(event_type, event_data):
...
+                                    [queue]- (service Y handler-meth)
+
+            requeue_on_error : bool  # TODO: defined by Consumer actually..
+                If true, handlers will return the event to the queue if an
+                error occurs while handling it. Defaults to False.
+            reliable_delivery : bool
+                Is true, events will be held in the queue until there is a
+                handler to consume them. Defaults to True.
+            sensitive_variables : string or tuple of strings
+                Mark an argument or part of an argument as sensitive. Stashed
```
Comment (davidszotten) [inline nameko/events.py]: "stashed" sounds a bit colloquial.

File: nameko/events.py [hunk 2]
```diff
@@ -101,64 +101,77 @@ def dispatch(event_type, event_data):
...
+                                  \
+                                    [queue]- (service Y handler-meth)
+
+            requeue_on_error : bool  # TODO: defined by Consumer actually..
+                If true, handlers will return the event to the queue if an
+                error occurs while handling it. Defaults to False.
+            reliable_delivery : bool
+                Is true, events will be held in the queue until there is a
+                handler to consume them. Defaults to True.
+            sensitive_variables : string or tuple of strings
```
Comment (davidszotten) [inline nameko/events.py]: do we definitely want to allow both a single or a tuple of strings?

File: nameko/utils.py [hunk 3]
```diff
@@ -1,9 +1,108 @@
...
+    method = getattr(entrypoint.container.service_cls, entrypoint.method_name)
+    argspec = inspect.getcallargs(method, None, *args, **kwargs)
+    del argspec['self']
+
+    def redact(data, keys):
+        key = keys[0]
+        if len(keys) == 1:
+            try:
+                data[key] = REDACTED
+            except (KeyError, IndexError):
```
Comment (davidszotten) [inline nameko/utils.py]: you can also get `TypeError` here if the argument doesn't match (e.g. trying to redact an integer thinking it's a dict)

File: test/test_utils.py [hunk 4]
```diff
@@ -50,3 +54,107 @@ def test_repr_safe_str(value, repr_safe_value):
     res = repr_safe_str(value)
     assert res == repr_safe_value
     assert isinstance(res, bytes)
+
+
+class TestGetRedactedArgs(object):
+
+    @pytest.mark.parametrize("sensitive_variables, expected", [
+        (tuple(), {'a': 'A', 'b': 'B'}),  # no sensitive variables
```
Comment (davidszotten) [inline test/test_utils.py]: `()` is an empty tuple

File: nameko/events.py (hunk 2 above)
Comment (mattbennett) [inline nameko/events.py]: Perhaps not. I was matching the interface of ``expected_exceptions``.  A common usecase would be something like:

``` python
class Service(object):
    @rpc(sensitive_variables=("foo",), expected_exceptions=FooError)
    def method(self, foo, bar):
        pass
```

The parenthesis seem slightly superfluous given we don't require them for ``expected_exceptions``.

File: nameko/utils.py (hunk 3 above)
Comment (mattbennett) [inline nameko/utils.py]: Good catch

File: test/test_utils.py (hunk 4 above)
Comment (mattbennett) [inline test/test_utils.py]: I find this more obvious -- easier to distinguish it amongst all the other parenthesis. Happy to change if you'd prefer though.

**Comment by davidszotten** [comment]
:+1: though we might want to hold of on landing this as part of 2.0

**Comment by mattbennett** [comment]
Re-raised against master -- https://github.com/onefinestay/nameko/pull/253

**Pull request closed without merging**